pkg/ruler/ruler.go

```diff
@@ -602,6 +606,93 @@ func filterRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, ring r
 	return result
 }
 
+// filterRuleGroupsByEnabled filters out from the input configs all the recording and/or alerting rules whose evaluation
+// has been disabled for the given tenant. This function modified the input configs in place.
+func filterRuleGroupsByEnabled(configs map[string]rulespb.RuleGroupList, limits RulesLimits, logger log.Logger) {
+	for userID, groups := range configs {
+		// Nothing to do if both recording and alerting rules are enabled.
+		recordingEnabled := limits.RulerRecordingRulesEvaluationEnabled(userID)
+		alertingEnabled := limits.RulerAlertingRulesEvaluationEnabled(userID)
+		if recordingEnabled && alertingEnabled {
+			continue
+		}
+
+		// Optimization in case both recording and alerting rules are disabled.
+		if !recordingEnabled && !alertingEnabled {
+			delete(configs, userID)
+			continue
+		}
+
+		var (
+			removedGroups     = 0
+			removedRulesTotal = 0
+		)
+
+		for groupIdx, group := range groups {
+			removedRules := 0
+
+			// Set to nil all rules that should be removed.
+			for ruleIdx, rule := range group.Rules {
+				if !alertingEnabled && rule.Alert != "" {
+					group.Rules[ruleIdx] = nil
+					removedRules++
+				}
+
+				if !recordingEnabled && rule.Record != "" {
+					group.Rules[ruleIdx] = nil
+					removedRules++
+				}
+			}
+
+			// If all rules were removed then we should the whole group,
+			// otherwise we'll just remove the rules set to nil.
+			if removedRules == len(group.Rules) {
+				groups[groupIdx] = nil
+				removedGroups++
+			} else if removedRules > 0 {
+				src, dst := 0, 0
+				for ; src < len(group.Rules); src++ {
+					if group.Rules[src] != nil {
+						group.Rules[dst] = group.Rules[src]
+						dst++
+					}
+				}
+
+				group.Rules = group.Rules[:dst]
+			}
+
+			removedRulesTotal += removedRules
+		}
+
+		// If all groups were removed then we should delete the user config,
+		// otherwise we'll just remove the groups set to nil.
+		if removedGroups == len(groups) {
+			delete(configs, userID)
+		} else if removedGroups > 0 {
+			src, dst := 0, 0
+			for ; src < len(groups); src++ {
+				if groups[src] != nil {
+					groups[dst] = groups[src]
+					dst++
+				}
+			}
+
+			configs[userID] = groups[:dst]
+		}
+
+		if removedRulesTotal > 0 {
+			// We don't expect rules evaluation to be disabled for the normal use case. For this reason,
```

Comment:
If we expect this to be rare does it make sense to return a new value in place instead of modifying the existing one to make the code a bit more straight forward? I know you mention avoiding any new allocations but it seems like we should prioritize clarity if this is indeed rare and would only be used during migrations.

Reply:
Yeah, I share your concern too and I can't take a decision myself :) It's not common to use it, but if you use it in a Mimir cluster with a large number of tenants / rules, then there could be a bit of impact if we reallocate all data structures instead of doing a in-place editing.

Reply:
I went for the copy version. See: https://github.com/grafana/mimir/pull/3088#issuecomment-1263352748